test/py/test_gpu_async.py

```diff
@@ -0,0 +1,78 @@
+#####################################################################################
+# The MIT License (MIT)
+#
+# Copyright (c) 2015-2022 Advanced Micro Devices, Inc. All rights reserved.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#####################################################################################
+import migraphx
+import ctypes
+
+
+def test_conv_relu():
+    libname = "libamdhip64.so"
+    hip = ctypes.cdll.LoadLibrary(libname)
+
+    p = migraphx.parse_onnx("conv_relu_maxpool_test.onnx")
+    print(p)
+    print("Compiling ...")
+    # Need to have offload_copy = False to avoid syncs() back to the host device
+    p.compile(migraphx.get_target("gpu"), offload_copy=False)
+    print(p)
+    params = {}
+
+    # Done to avoid parsing enums in ctypes. hipGetDevice always gives us
+    # hipSuccess as an output without modifying state of the current device
+    device_id = ctypes.c_void_p()
+    hipSuccess = ctypes.c_long(hip.hipGetDevice(device_id))
+    migraphx.gpu_sync()
+
+    # Alloc a stream
+    stream = ctypes.c_void_p()
+
+    err = ctypes.c_long(
+        hip.hipStreamCreateWithFlags(ctypes.addressof(stream), 0))
+
+    if err != hipSuccess:
+        print("hipStreamCreate failed")
+        return err
+
+    # Use to_gpu to push generated argument to the GPU before we perform a run
+    for key, value in p.get_parameter_shapes().items():
+        params[key] = migraphx.to_gpu(migraphx.generate_argument(value))
+
+    result = p.run_async(params, stream, "ihipStream_t")
+
+    # Wait for all commands in stream to complete
+    err = ctypes.c_long(hip.hipStreamSyncrhonize(stream))
```

Comment:
Odd I have this in the CMakeLists.txt file for this change

Reply:
The usual result I see is
289/305 Test #305: test_py_3.8_gpu_async .....................................................   Passed   11.58 sec

Oddly if I keep running make -j$(nproc) check over and over I can cause this.


288/305 Test #305: test_py_3.8_gpu_async .....................................................***Failed   10.33 sec
Segmentation fault

Something non obvious is happening,  I wasn't seeing any output when I ran with the 

```
if __name__ == "main":
```
at the start to run the test_conv_relu

Reply:
Soo bit of a fun one and a lot of non obvious things where.

```
a = ctypes.<some standard type>
b = ctypes.<some standard type>


print (a == b)   -> false.   # Fails since we need to compare value attributes
print(a.value == b.value) # True
```
This explains why sometimes some of the calls would fail, silently even though passing

Combined with how we pass by reference:
```
a = ctypes.c_void_p()
result = c_func(addressof(a))  # causes undefined behavior

#alternative
result = c_func(byref(a)) # works
```
Turns out ctypes has some overhead with both addressof() and byref() calls that handle pass by reference correctly. 

Removed these and then ran into the spelling error around with ` hipStreamSyncrhonize` finally triggering.

Bit of an onion but this runs all the way through now. Just for some reason using the output of run_async in a print results in a segfault still.